File: server/src/main/java/org/codi/catan/impl/game/MoveApiHelper.java
/*
 * @author the-codinator
 * created on 2020/5/30
 */

package org.codi.catan.impl.game;

import javax.inject.Inject;
import javax.inject.Singleton;
import javax.ws.rs.core.Response.Status;
import org.codi.catan.core.CatanException;
import org.codi.catan.impl.data.CatanDataConnector;
import org.codi.catan.impl.handler.BoardStateColorRequestMoveHandler;
import org.codi.catan.impl.handler.BoardStateMoveHandler;
import org.codi.catan.impl.handler.BoardStateRequestMoveHandler;
import org.codi.catan.impl.handler.StateColorRequestMoveHandler;
import org.codi.catan.impl.handler.StateMoveHandler;
import org.codi.catan.impl.handler.StateRequestMoveHandler;
import org.codi.catan.model.game.Board;
import org.codi.catan.model.game.Color;
import org.codi.catan.model.game.OutOfTurnApi;
import org.codi.catan.model.game.Phase;
import org.codi.catan.model.game.State;
import org.codi.catan.model.response.StateResponse;
import org.codi.catan.util.Util;

@Singleton
public class MoveApiHelper {

    private final BoardHelper boardHelper;
    private final StateApiHelper stateApiHelper;
    private final GameUtility gameUtility;
    private final CatanDataConnector dataConnector;

    @Inject
    public MoveApiHelper(BoardHelper boardHelper, StateApiHelper stateApiHelper, GameUtility gameUtility,
        CatanDataConnector dataConnector) {
        this.boardHelper = boardHelper;
        this.stateApiHelper = stateApiHelper;
        this.gameUtility = gameUtility;
        this.dataConnector = dataConnector;
    }

    /**
     * Generic function to play a move
     */
    public <T> StateResponse play(OutOfTurnApi outOfTurnApi, String userId, String gameId, String etag, T request,
        BoardStateColorRequestMoveHandler<T> handler, Phase... validPhases) throws CatanException {
        Board board = boardHelper.getBoard(gameId);
        State state = stateApiHelper.getState(gameId, null);
        if (etag != null && !state.getETag().equals(etag)) {
            throw new CatanException("Someone made a move before you. Please retry with updated game state.",
                Status.PRECONDITION_FAILED);
        }
        Color color = gameUtility.checkPlayerTurn(board, state, userId, outOfTurnApi);
        gameUtility.ensurePhaseForMove(state, validPhases);
        if (handler.shouldValidateInput()) {
            Util.validateInput(request);
        }
        handler.play(board, state, color, request);
        try {
            dataConnector.updateState(state);
        } catch (CatanException e) {
            if (e.getErrorStatus() == Status.PRECONDITION_FAILED) {
                throw new CatanException("Someone made a move before you. Please retry with updated game state.",
                    Status.PRECONDITION_FAILED, e);
            } else {
                throw e;
            }
        }
        return new StateResponse(state, color);
    }

    public <T> StateResponse play(OutOfTurnApi outOfTurnApi, String userId, String gameId, String etag, T request,
        StateColorRequestMoveHandler<T> handler, Phase... validPhases) throws CatanException {
        return play(outOfTurnApi, userId, gameId, etag, request, handler.asBaseType(), validPhases);
    }

    public <T> StateResponse play(String userId, String gameId, String etag, T request,
        BoardStateRequestMoveHandler<T> handler, Phase... validPhases) throws CatanException {
        return play(null, userId, gameId, etag, request, handler.asBaseType(), validPhases);
    }

    public StateResponse play(String userId, String gameId, String etag, BoardStateMoveHandler handler,
        Phase... validPhases) throws CatanException {
        return play(null, userId, gameId, etag, null, handler.asBaseType(), validPhases);
    }

    public <T> StateResponse play(String userId, String gameId, String etag, T request,
        StateRequestMoveHandler<T> handler, Phase... validPhases) throws CatanException {
        return play(null, userId, gameId, etag, request, handler.asBaseType(), validPhases);
    }

    public StateResponse play(String userId, String gameId, String etag, StateMoveHandler handler, Phase... validPhases)
        throws CatanException {
        return play(null, userId, gameId, etag, null, handler.asBaseType(), validPhases);
    }
}

File: server-azure/src/impl/game/move-api-helper.ts
import type { BodyLessRequest, MoveRequest } from '../../model/request';
import { StateResponse, createStateResponse } from '../../model/response/state-response';
import { UserGame, createUserGame, createUserGamesFromBoard } from '../../model/game/user-game';
import { checkPlayerTurn, ensurePhaseForMove } from './game-utility';
import type { Board } from '../../model/game/board';
import type { CatanContext } from '../../core/catan-context';
import { CatanError } from '../../core/catan-error';
import type { Color } from '../../model/game/color';
import type { DeepReadonly } from 'ts-essentials';
import type { OutOfTurnApi } from '../../model/game/out-of-turn-api';
import { PRECONDITION_FAILED } from 'http-status-codes';
import { Phase } from '../../model/game/phase';
import type { State } from '../../model/game/state';
import type { StrongEntity } from '../../model/core';
import type { Writable } from 'ts-essentials';
import dataConnector from '../data/catan-data-connector';
import { getBoard } from './board-helper';
import { getState } from './state-api-helper';

export interface PlayOptions<T extends MoveRequest> {
  board: DeepReadonly<Board>;
  state: State & Readonly<StrongEntity>;
  color: Readonly<Color>;
  request: DeepReadonly<T extends BodyLessRequest ? undefined : T>;
}

export async function play<T extends MoveRequest>(
  outOfTurnApi: OutOfTurnApi | undefined,
  context: Pick<CatanContext<T>, 'user' | 'gameId' | 'etag' | 'request'>,
  handler: (options: PlayOptions<T>) => void,
  ...validPhases: Phase[]
): Promise<StateResponse> {
  const { user, gameId, etag, request } = context;
  const board = await getBoard(gameId);
  const state = await getState(gameId, undefined);
  const current = state.currentMove.color;
  if (etag && etag !== state.etag) {
    throw new CatanError('Someone made a move before you. Please retry with updated game state.', PRECONDITION_FAILED);
  }
  const color = checkPlayerTurn(board, state, user, outOfTurnApi);
  ensurePhaseForMove(state, validPhases);
  const options: PlayOptions<T> = { board, state, color, request };
  handler(options); // Mutates state
  try {
    await dataConnector.updateState(state);
    const updated = state.currentMove.color;
    if (updated !== current) {
      const oldUG = createUserGame(board, current);
      const newUG = createUserGame(board, updated);
      (newUG as Writable<UserGame>).myTurn = true;
      await dataConnector.updateUserGames(oldUG, newUG);
    } else if (state.phase === Phase.end && board.completed === 0) {
      board.completed = Date.now();
      await dataConnector.updateBoard(board);
      const ugs = createUserGamesFromBoard(board);
      await dataConnector.updateUserGames(...ugs);
    }
  } catch (e) {
    if (e.errorState === PRECONDITION_FAILED) {
      throw new CatanError(
        'Someone made a move before you. Please retry with updated game state.',
        PRECONDITION_FAILED,
        e
      );
    } else {
      throw e;
    }
  }
  return createStateResponse(state, color);
}
